File: xmlBlaster/testsuite/src/c++/TestCommand.cpp
/*-----------------------------------------------------------------------------
Name:      TestCommand.cpp
Project:   xmlBlaster.org
Copyright: xmlBlaster.org, see xmlBlaster-LICENSE file
Comment:   Demo code for a client using xmlBlaster
Version:   $Id: TestCommand.cpp 12915 2004-11-18 14:55:44Z ruff $
-----------------------------------------------------------------------------*/
#include "TestSuite.h"
#include <iostream>

using namespace std;
using namespace org::xmlBlaster::util;
using namespace org::xmlBlaster::util::qos;
using namespace org::xmlBlaster::util::thread;
using namespace org::xmlBlaster::client;
using namespace org::xmlBlaster::client::qos;
using namespace org::xmlBlaster::client::key;
using namespace org::xmlBlaster::authentication;

namespace org { namespace xmlBlaster { namespace test {

/**
 * This client tests the method sendAdministrativeCommand(). 
 * This client may be invoked multiple time on the same xmlBlaster server,
 * as it cleans up everything after his tests are done.
 */
class SpecificCallback : public I_Callback {
private:
   int numReceived_;
   string name_;
   I_Log& log_;

public:
   SpecificCallback(I_Log& log, const string& name) : log_(log) {
      name_ = name;
      numReceived_ = 0;
   }

   int getCount() {
      return numReceived_;
   }


   string update(const string &sessionId,
               UpdateKey &updateKey,
               const unsigned char * /*content*/, long /*contentSize*/,
               UpdateQos &updateQos) 
   {
      log_.info("update", string("Receiving update on callback '") + name_ + "' of message oid=" +
                updateKey.getOid() + " state=" + updateQos.getState() +
                " authentication sessionId=" + sessionId + " ...");
      numReceived_++;
      return "<qos><state id='OK'/></qos>";
   }


};


class TestCommand: public TestSuite, public virtual I_Callback 
{
private:
   bool   messageArrived_;      // = false;
   int    numReceived_;         //  = 0;         // error checking
   string subscribeOid_;
   string publishOid_;          // = "dummy";
   string senderName_;
   string senderContent_;
   string receiverName_;        // sender/receiver is here the same client
   string contentMime_;         // = "text/xml";
   string contentMimeExtended_; //  = "1.0";
   ConnectReturnQos returnQos_;
   SpecificCallback *cb1_;
   SpecificCallback *cb2_;
   SpecificCallback *cb3_;

   /** Publish tests */
   enum TestType {
      TEST_ONEWAY, TEST_PUBLISH, TEST_ARRAY
   };

   /**
    * Constructs the TestCommand object.
    * <p />
    * @param testName  The name used in the test suite
    * @param loginName The name to login to the xmlBlaster
    */
 public:
   TestCommand(int args, char *argc[], const string &loginName)
      : TestSuite(args, argc, "TestCommand"), returnQos_(global_)
   {
      senderName_          = loginName;
      receiverName_        = loginName;
      numReceived_         = 0;
      publishOid_          = "dummy";
      contentMime_         = "text/xml";
      contentMimeExtended_ = "1.0";
      senderContent_       = "Yeahh, i'm the new content";
      cb1_ = new SpecificCallback(log_, "callback1");
   }

   virtual ~TestCommand() 
   {
      delete cb1_;
   }

   /**
    * Sets up the fixture. <p />
    * Connect to xmlBlaster and login
    */
   void setUp() 
   {
      log_.info(ME, "Trying to connect to xmlBlaster with C++ client lib " + Global::getVersion() + " from " + Global::getBuildTimestamp());
      TestSuite::setUp();
      try {
         string passwd = "secret";
         SecurityQos secQos(global_, senderName_, passwd);
         ConnectQos connQos(global_);
         connQos.getSessionQosRef()->setPubSessionId(3L);
         returnQos_ = connection_.connect(connQos, this);
         string name = returnQos_.getSessionQos().getAbsoluteName();
         string name1 = returnQos_.getSessionQosRef()->getAbsoluteName();
         assertEquals(log_, ME, name, name1, string("name comparison for reference"));

         log_.info(ME, string("connection setup: the session name is '") + name + "'");
         // Login to xmlBlaster
      }
      catch (XmlBlasterException &e) {
         log_.error(ME, string("Login failed: ") + e.toXml());
         usage();
         assert(0);
      }
   }


   /**
    * Tears down the fixture. <p />
    * cleaning up .... erase() the previous message OID and logout
    */
   void tearDown() 
   {
      log_.info(ME, "Cleaning up test - erasing message.");
      /*
      EraseKey eraseKey(global_);
      eraseKey.setOid(publishOid_);
      EraseQos eraseQos(global_);

      vector<EraseReturnQos> retArr;
      try {
         retArr = connection_.erase(eraseKey, eraseQos);
      }
      catch(XmlBlasterException &e) {
         log_.error(ME, string("XmlBlasterException: ") + e.toXml());
      }
      if (retArr.size() != 1) {
         log_.error(ME, "Erased " + lexical_cast<string>(retArr.size()) + " messages");
      }
      */
      connection_.disconnect(DisconnectQos(global_));
      TestSuite::tearDown();
   }


   void testSetCallbackDispatcherActive() 
   {
      if (log_.trace()) log_.trace(ME, "setCallbackDispatcherActive() ...");
      try {
         connection_.setCallbackDispatcherActive(false);
         log_.info(ME, string("Success: setCallbackDispatcherActive(false)"));

         //string ret = org::xmlBlaster::util::waitOnKeyboardHit("Hit a key to activate again ...");

         connection_.setCallbackDispatcherActive(true);
         log_.info(ME, string("Success: setCallbackDispatcherActive(true)"));

         //org::xmlBlaster::util::waitOnKeyboardHit("Hit a key to finish test ...");
      }
      catch(XmlBlasterException &e) {
         log_.warn(ME, string("XmlBlasterException: ") + e.toXml());
         assert(0);
      }
   }


   void testSendAdministrativeCommand() 
   {
      if (log_.trace()) log_.trace(ME, "sendAdministrativeCommand() ...");
      //string command = "?clientList";
      {
         string command = global_.getId()+"/?dispatcherActive=false";
         log_.info(ME, string("Trying command '" + command + "'"));
         try {
            string ret = connection_.sendAdministrativeCommand(command);
            log_.info(ME, string("Success: " + command + " returned '" + ret + "'"));
         }
         catch(XmlBlasterException &e) {
            log_.warn(ME, string("XmlBlasterException: ") + e.toXml());
            assert(0);
         }
      }
      {
         string command = global_.getId()+"/?dispatcherActive";
         log_.info(ME, string("Trying command '" + command + "'"));
         try {
            string ret = connection_.sendAdministrativeCommand(command);
            log_.info(ME, string("Success: " + command + " returned '" + ret + "'"));
         }
         catch(XmlBlasterException &e) {
            log_.warn(ME, string("XmlBlasterException: ") + e.toXml());
            assert(0);
         }
      }
   }

   string update(const string &sessionId,
               UpdateKey &updateKey,
               const unsigned char *content, long contentSize,
               UpdateQos &updateQos) 
   {
      log_.info(ME, string("Receiving update of message oid=") +
                updateKey.getOid() + " state=" + updateQos.getState() +
                " authentication sessionId=" + sessionId + " ...");
      numReceived_ ++;
      return "<qos><state id='OK'/></qos>";
   }
};

}}} // namespace

using namespace org::xmlBlaster::test;

int main(int args, char *argc[]) 
{
   try {
      org::xmlBlaster::util::Object_Lifetime_Manager::init();
      TestCommand testSub(args, argc, "Tim");
 
      testSub.setUp();
      testSub.testSetCallbackDispatcherActive();
      testSub.testSendAdministrativeCommand();
      testSub.tearDown();

      Thread::sleepSecs(1);
   }
   catch (XmlBlasterException& ex) {
      std::cout << ex.toXml() << std::endl;
   }
   catch (...) {
      cout << "unknown exception occured" << endl;
      XmlBlasterException e(INTERNAL_UNKNOWN, "main", "main thread");
      cout << e.toXml() << endl;
   }
   org::xmlBlaster::util::Object_Lifetime_Manager::fini();
   return 0;
}

File: testsuite/src/java/org/xmlBlaster/test/client/TestCommand.java
package org.xmlBlaster.test.client;

import java.util.logging.Logger;

import org.xmlBlaster.client.I_XmlBlasterAccess;
import org.xmlBlaster.client.key.UpdateKey;
import org.xmlBlaster.client.qos.ConnectQos;
import org.xmlBlaster.client.qos.DisconnectQos;
import org.xmlBlaster.client.qos.UpdateQos;
import org.xmlBlaster.test.MsgInterceptor;
import org.xmlBlaster.util.Global;
import org.xmlBlaster.util.XmlBlasterException;

import junit.framework.TestCase;

/**
 * Testing administrative commands. 
 * @author xmlblast@marcelruff.info
 */
public class TestCommand extends TestCase {
   private final Global global;
   private static Logger log = Logger.getLogger(TestCommand.class.getName());
   private Global connGlobal;
   private MsgInterceptor updateInterceptor;

   public static void main(String[] args) {
      junit.swingui.TestRunner.run(TestCommand.class);
   }

   /**
    * Constructor for TestCommand.
    * @param arg0
    */
   public TestCommand(String name) {
      super(name);
      this.global = Global.instance();
   }

   /*
    * @see TestCase#setUp()
    */
   protected void setUp() throws Exception {
      super.setUp();
      log.info("Trying to connect to xmlBlaster with Java client lib " + global.getVersion());
      try {
         this.connGlobal = this.global.getClone(null);
         this.updateInterceptor = new MsgInterceptor(this.connGlobal, log, null);
         this.connGlobal.getXmlBlasterAccess().connect(new ConnectQos(this.connGlobal), this.updateInterceptor);
      }
      catch (XmlBlasterException ex) {
         ex.printStackTrace();
         fail("aborting since exception ex: " + ex.getMessage());
      }
   }

   /*
    * @see TestCase#tearDown()
    */
   protected void tearDown() throws Exception {
      super.tearDown();
      log.info("Entering tearDown(), test is finished");
      this.connGlobal.getXmlBlasterAccess().disconnect(new DisconnectQos(this.connGlobal));
      this.connGlobal.shutdown();
      this.connGlobal = null;
   }

   public String update(String cbSessionId, UpdateKey updateKey, byte[] content, UpdateQos updateQos) throws XmlBlasterException {
      log.severe("update: should never be invoked (msgInterceptors take care of it since they are passed on subscriptions)");
      return "OK";
   }
   
   public void testSetCallbackDispatcherActive() {
      log.fine("setCallbackDispatcherActive() ...");
      try {
         this.connGlobal.getXmlBlasterAccess().setCallbackDispatcherActive(false);
         log.info("Success: setCallbackDispatcherActive(false)");

         //Global.waitOnKeyboardHit("Hit a key to activate again ...");

         this.connGlobal.getXmlBlasterAccess().setCallbackDispatcherActive(true);
         log.info("Success: setCallbackDispatcherActive(true)");

         //Global.waitOnKeyboardHit("Hit a key to finish test ...");
      }
      catch(XmlBlasterException e) {
         log.warning("XmlBlasterException: " + e.toXml());
         fail(e.getMessage());
      }
   }

   public void testSendAdministrativeCommand() {
      log.fine("sendAdministrativeCommand() ...");
      I_XmlBlasterAccess con = this.connGlobal.getXmlBlasterAccess();
      String sessionName = con.getSessionName().getAbsoluteName();
      {
         String command = sessionName+"/?dispatcherActive";
         log.info("Trying command '" + command + "'");
         try {
            String ret = con.sendAdministrativeCommand(command);
            log.info("Success: " + command + " returned '" + ret + "'");
            assertEquals("true", ret);
         }
         catch(XmlBlasterException e) {
            fail(e.getMessage());
         }
      }
      {
         String command = sessionName+"/?dispatcherActive=false";
         log.info("Trying command '" + command + "'");
         try {
            String ret = con.sendAdministrativeCommand(command);
            log.info("Success: " + command + " returned '" + ret + "'");
            assertEquals("OK", ret);
         }
         catch(XmlBlasterException e) {
            fail(e.getMessage());
         }
      }
      {
         String command = sessionName+"/?dispatcherActive";
         log.info("Trying command '" + command + "'");
         try {
            String ret = con.sendAdministrativeCommand(command);
            log.info("Success: " + command + " returned '" + ret + "'");
            assertEquals("false", ret);
         }
         catch(XmlBlasterException e) {
            fail(e.getMessage());
         }
      }
      {
         String command = "set " + sessionName+"/?dispatcherActive=true";
         log.info("Trying command '" + command + "'");
         try {
            String ret = con.sendAdministrativeCommand(command);
            log.info("Success: " + command + " returned '" + ret + "'");
            assertEquals("OK", ret);
         }
         catch(XmlBlasterException e) {
            fail(e.getMessage());
         }
      }
      {
         String command = "get " + sessionName+"/?dispatcherActive";
         log.info("Trying command '" + command + "'");
         try {
            String ret = con.sendAdministrativeCommand(command);
            log.info("Success: " + command + " returned '" + ret + "'");
            assertEquals("true", ret);
         }
         catch(XmlBlasterException e) {
            fail(e.getMessage());
         }
      }
   }
}
